### PR TITLE
[RLlib; Offline RL] Add `CQLLearner` and `CQLTorchLearner`.

### LINKED ISSUE #37779: [RLlib] Enabling RLModule by default on CQL
### Description

We need to enable the new stack by default. 

### Use case

_No response_

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

This PR proposes the `Learner` classes for `CQL` in the new API stack. It relies on the `SACTorchModule` (i.e. having a Q-network and Q-target-network) and implements the loss logic. Because three loss terms are optimized the gradients are collected inside the `compute_loss_for_module` method and are then just returned from `compute_gradients`. The reason for this are multiple passes through the networks where a follow-up path would erase the gradients computed in the pass before.

This PR is one of several PRs that should implement `CQL` in the new API stack. 

## Related issue number

Closes #37779

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/cql/torch/cql_torch_learner.py [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+        # TODO (simon): Check, why log(alpha) is used, prob. just better
+        # to optimize and monotonic function. Original equation uses alpha.
+        alpha_loss = -torch.mean(
+            self.curr_log_alpha[module_id]
+            * (logps_curr.detach() + self.target_entropy[module_id])
+        )
+
+        # Get the current batch size. Note, this size might vary in case the
+        # last batch contains less than `train_batch_size_per_learner` examples.
+        batch_size = batch[Columns.OBS].shape[0]
```
Comment (sven1977) [inline rllib/algorithms/cql/torch/cql_torch_learner.py:81]: Can we explain this logic here and why we defend against different batch sizes coming in?

Don't we expect always the same batch size from the data pipeline?

If not, we should:
a) explain here why we are expecting various batch sizes
b) probably fix this logic here. What if we call `compute_loss_for_module` 10x with a batch of `train_batch_size_per_learner - 1` (accumulating gradients for these, but not applying these) and then one batch of size `train_batch_size_per_learner`. In this case, the effective batch size would be roughly 11x the user configured one, correct?

c) Also, what if the incoming batch is larger than `train_batch_size_per_learner`?

File: rllib/algorithms/cql/cql_learner.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class CQLLearner(SACLearner):
+    @override(Learner)
+    def build(self) -> None:
+        # We need to call the `super()`'s `build` method here to have the variables
+        # for `alpha`` and the target entropy defined.
+        super().build()
+
+        # Set up the gradient buffer to store gradients to apply
```
Comment (sven1977) [inline rllib/algorithms/cql/cql_learner.py:15]: Can we add more explanation here as to why caching any grads is necessary?

Maybe I don't understand the data pipelines properly yet, but should we try to fix these to always produce the exact same batch sizes (or number of episode timesteps) per iteration?

File: rllib/algorithms/cql/torch/cql_torch_learner.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/algorithms/cql/torch/cql_torch_learner.py:81]: @sven1977 I fully agree on your comment. 

a) This was also part of the old stack algorithm. The reason behind this is that when iterating over a dataset the last batch could have less than `train_batch_size_per_learner` samples in it. 

b) The offline logic here is that after each call to `compute_loss_for_module` a `compute_gradients` and `apply_gradients` will occur. No SGD is run on the offline algorithms (yet). So the train batch size should be always as large as configured (without the last one being smaller). 

c) This case should also not happen. `iter_batches` takes care of this and ensures that always the batch size is sampled, without the last one which can be avoided by setting a flag, but neglects data.

File: rllib/algorithms/cql/cql_learner.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/algorithms/cql/cql_learner.py:15]: This is not really caused by the batch sizes, but will only be executed if the batch size is as configured. The reason for the grad caching is merely to enable multiple passes through the network during loss calculation.

File: rllib/algorithms/cql/cql_learner.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        # them later in `self.apply_gradients`.
+        self.grads = {}
+
+        # Log the training iteration to switch from behavior cloning to improving
+        # the policy.
+        # TODO (simon, sven): Add upstream information pieces into this timesteps
+        # call arg to Learner.update_...().
+        self.metrics.log_value(
```
Comment (sven1977) [inline rllib/algorithms/cql/cql_learner.py:23]: Add: reduce="sum" here to make sure iterations are not EMA'd :) (the default).

File: rllib/algorithms/cql/torch/cql_torch_learner.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+        *,
+        module_id: ModuleID,
+        config: CQLConfig,
+        batch: Dict,
+        fwd_out: Dict[str, TensorType],
+    ) -> TensorType:
+
+        # TODO (simon, sven): Add upstream information pieces into this timesteps
+        # call arg to Learner.update_...().
+        self.metrics.log_value(
```
Comment (sven1977) [inline rllib/algorithms/cql/torch/cql_torch_learner.py:43]: New that reduce="sum" is already defined above, you can just do:

```
log_value(
    (ALL_MODULES, TRAINING_ITERATION),
    1,
)
```

File: rllib/algorithms/cql/cql_learner.py (hunk 3 above)
Comment (sven1977) [reply rllib/algorithms/cql/cql_learner.py:23]: Actually, you don't need this here anymore. B/c you are logging it below, this entry will be created automatically.

File: rllib/algorithms/cql/torch/cql_torch_learner.py [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+                        self._params, optim
+                    ).items()
+                }
+            )
+
+        # Get the current alpha.
+        alpha = torch.exp(self.curr_log_alpha[module_id])
+        # Start training with behavior cloning and turn to the classic Soft-Actor Critic
+        # after `bc_iters` of training iterations.
+        if self.metrics.peek((ALL_MODULES, TRAINING_ITERATION)) >= config.bc_iters:
```
Comment (sven1977) [inline rllib/algorithms/cql/torch/cql_torch_learner.py:104]: Change to:
```
self.metrics.peek(..., default=0)
```

**Review by sven1977 — APPROVED** [review]
Just some nits left, otherwise good to go! Awesome work @simonsays1980 .
